README.md
senora
======

Senora is targeted to be a drop-in replacement for sqlplus. It provides a plethora of new commands, which accept Unix-style options. It is enhanceable via plug-ins, and provides most of sqlplus' functionality.

Senora intends to be your primary Oracle shell. While there are may nifty Oracle tools around (e.g. TORA or TOAD), these do not attempt to replace sqlplus, so chances are you'll need to run sqlplus additionally. Also these tools are hard (TORA) or impossible (TOAD, sqlplus) to extend. Senora attempts to give you much of the browsing and anayzing capabilities of these tools without the need to run any other tool. 

You can extend Senora easily by providing you own plugins. In fact most of Senora's core functionality is written as plugins. A plugin provides additional commands to Senora and integrates these with Senora's help system. Senora's extensions are thus self-documenting as we know it from emacs.

This is just one benefit of plugins over stand-alone sql-scripts. Another one is the ability to provide unix-style options. The lack of options in sqlplus was in fact the main reason I started to write senora. Options allows you to group similar commands under one common hood, even when the undelying SQL differs significantly depending on the currently set options. The good thing about this is: you need to remeber less, and if things do slip your memory, the help system will guide you - much faster than any menu driven system ever will. 

Some Plugins have grown into extremely useful tools. I am getting a reputation for being an oracle wizard, just becuase I fire up the Senora's profiler (from the Tuning plugin) whenver someone ask the question "why is this so slow ?"

Senora attempts to provide a friendlier output formatting than sqlplus. Columns tend to be only as wide a really needed, and when linebreaks are needed, these are placed at "good" positions, i.e. after blanks or undescores. This is of course an incompatibility to sqlplus, and programs that rely on sqlplus-style fromatting may not work with Senora. Still Senora understands some of sqlpluses columns formatting commands.

Senora can run legacy sqlplus scripts as long as these do non use any unsupported commands. The most important sqlplus commands however do work, including starting scripts from within a script (@ and @@), using defines (see ampersand replacing) and bind variables.
 
	


changes.txt
----------------------------
revision 1.28
date: 2001/05/22 19:47:59;  author: drautzburg;  state: Exp;  lines: +1 -0

Fixed serveroutput
----------------------------
revision 1.27
date: 2001/05/22 19:29:51;  author: drautzburg;  state: Exp;  lines: +1 -0

more Prompt issues. Handle lines without newlines at the end of scripts.
----------------------------
revision 1.26
date: 2001/05/22 17:55:54;  author: drautzburg;  state: Exp;  lines: +1 -0

Better prompting: no prompt when processing file. Hopefully works now.
----------------------------
revision 1.25
date: 2001/05/21 19:28:02;  author: drautzburg;  state: Exp;  lines: +1 -0

fixed typo in DataDictionary.pm
----------------------------
revision 1.24
date: 2001/05/21 18:17:39;  author: drautzburg;  state: Exp;  lines: +3 -0

Tuning.pm
    Added validate command
    Added rollSegs command
    Added ps command (processes)
    Added hwm command (high water mark)
    Added stat command (to print various stats)
    Added job command (to display jobs)
    Enhanced xqueries (to print most expensive queries)
    Added xplain command (to show execution plans)
    Added kept command (to show pinned code)
    Added space command (to show tablespace stats)
    Added waits command (to show what sessions are waitng for)
    Added locks command (to show locks and locked objects)
    Added show parameter
----------------------------
revision 1.23
date: 2001/05/21 18:09:59;  author: drautzburg;  state: Exp;  lines: +1 -0

made describe print the realname of an object in case it is a synonym
plus some more changes.
----------------------------
revision 1.22
date: 2001/05/21 18:08:09;  author: drautzburg;  state: Exp;  lines: +1 -0

Enhanced formatting of columns containing newlines: break at newline
if possible. Nice for queries agains sqltext_with_newlines
----------------------------
revision 1.21
date: 2001/05/21 18:06:05;  author: drautzburg;  state: Exp;  lines: +1 -0

Extend @INC so senora can be started from anywhere
----------------------------
revision 1.20
date: 2001/05/21 18:04:58;  author: drautzburg;  state: Exp;  lines: +1 -0

Lots of changes in Qs module. Qs should go from CVS actually
----------------------------
revision 1.19
date: 2001/05/21 18:02:07;  author: drautzburg;  state: Exp;  lines: +1 -0

added min function
----------------------------
revision 1.18
date: 2001/05/21 17:47:12;  author: drautzburg;  state: Exp;  lines: +1 -0

added rollback create\s+role revoke lock analyze + /*sql*/ hint
to force sql excution
----------------------------
revision 1.17
date: 2001/05/21 17:44:44;  author: drautzburg;  state: Exp;  lines: +1 -0

 tried to give more feedback. Still needs more work.
----------------------------
revision 1.16
date: 2001/05/21 17:43:25;  author: drautzburg;  state: Exp;  lines: +2 -0

FdStack: played with prompts. Not satisfied yet
----------------------------
revision 1.15
date: 2001/05/21 17:41:11;  author: drautzburg;  state: Exp;  lines: +1 -0


Added -c option to "ls" to list constraints and indexes and stuff
----------------------------
revision 1.14
date: 2001/05/02 18:28:51;  author: drautzburg;  state: Exp;  lines: +1 -0

Trying to improve Windows compatibilty. Startetd xplain in Tuning.pm
----------------------------
revision 1.13
date: 2001/04/30 02:30:27;  author: drautzburg;  state: Exp;  lines: +1 -0

Fixed bug when lines contain lonely semicolon
----------------------------
revision 1.12
date: 2001/04/30 01:24:33;  author: drautzburg;  state: Exp;  lines: +1 -0

Added support for bind variables including the new commands
VARIABLE and PRINT.
----------------------------
revision 1.11
date: 2001/04/28 14:07:08;  author: drautzburg;  state: Exp;  lines: +1 -0

Moved printError from SqlSession to Feedback


** SENORA 2.5

----------------------------
revision 1.10
date: 2001/04/28 13:31:21;  author: drautzburg;  state: Exp;  lines: +1 -0

updated README
----------------------------
revision 1.9
date: 2001/04/28 13:27:51;  author: drautzburg;  state: Exp;  lines: +1 -0

Fixed bug in reporting success of ALTER SESSION command.
----------------------------
revision 1.8
date: 2001/04/28 13:15:22;  author: drautzburg;  state: Exp;  lines: +1 -0

Fixed bug in set serveroutput: on/off status was incorrectly reported
when sessions were switched.
----------------------------
revision 1.7
date: 2001/04/28 13:07:00;  author: drautzburg;  state: Exp;  lines: +1 -0

Fixed ampersandReplace so the period is ignored in &foo.bar
----------------------------
revision 1.6
date: 2001/04/28 12:58:45;  author: drautzburg;  state: Exp;  lines: +4 -1

Added PROMPT command
----------------------------
revision 1.5
date: 2001/04/28 12:55:46;  author: drautzburg;  state: Exp;  lines: +2 -3
*** empty log message ***
----------------------------
revision 1.4
date: 2001/04/26 23:50:04;  author: drautzburg;  state: Exp;  lines: +2 -1

no comment
----------------------------
revision 1.3
date: 2001/04/26 23:31:26;  author: drautzburg;  state: Exp;  lines: +2 -4

Added REM and -- comments
----------------------------
revision 1.2
date: 2001/04/24 20:52:30;  author: drautzburg;  state: Exp;  lines: +2 -1

no comment
----------------------------
revision 1.1
date: 2001/03/13 18:58:40;  author: drautzburg;  state: Exp;
branches:  1.1.1;
Initial revision
----------------------------
revision 1.1.1.1
date: 2001/03/13 18:58:40;  author: drautzburg;  state: Exp;  lines: +0 -0

4th attempt of initial checkin
=============================================================================
martin@beaureve:~/projects/senora/senora$ 

Wed May 23 23:50:45 CEST 2001
Thu May 24 00:06:43 CEST 2001
Fri Dec 21 19:57:04 CET 2001
Sun Dec 30 01:02:49 CET 2001
Sun Dec 30 01:04:43 CET 2001
Sun Dec 30 01:05:23 CET 2001
Sun Dec 30 01:06:11 CET 2001
Mon Dec 31 21:34:54 CET 2001

core/DDL/ddl.pl
#! /usr/bin/perl -w

# $Id: ddl.pl,v 1.1 2011/03/25 14:36:26 ekicmust Exp $

use strict;

use DBI;
use DDL::Oracle;
use English;

my  $dbh = DBI->connect(
                        "dbi:Oracle:",
                        "",
                        "",
                        {
                         PrintError => 0,
                         RaiseError => 1
                        }
    );

DDL::Oracle->configure( 
                        dbh      => $dbh,
                        resize   => 1,
#                        view     => 'user',
#                        heading  => 0,
#                        prompt   => 0,
                      );

my $user = getlogin
        || scalar getpwuid($REAL_USER_ID)
        || undef;

print STDERR "Enter Action [CREATE]: ";
chomp( my $action = <STDIN> );
$action = "create" unless $action;

print STDERR "Enter Type    [TABLE]: ";
chomp( my $type = <STDIN> );
$type = "TABLE" unless $type;

print STDERR "Enter Owner [\U$user]: ";
chomp( my $owner = <STDIN> );
$owner = $user unless $owner;
die "\nYou must specify an Owner.\n" unless $owner;

print STDERR "Enter Name           : ";
chomp( my $name = <STDIN> );
die "\nYou must specify an object.\n"
   unless (
                $name
             or "\U$type" eq 'COMPONENTS'
             or "\U$type" eq 'SCHEMA'
          );

print STDERR "\n";

my $obj = DDL::Oracle->new(
                            type  => $type,
                            list  => [
                                       [
                                         $owner,
                                         $name,
                                       ]
                                     ]
                          );

my $sql;

if ( $action eq "drop" ){
    $sql = $obj->drop;
}
elsif ( $action eq "create" ){
    $sql = $obj->create;
}
elsif ( $action eq "resize" ){
    $sql = $obj->resize;
}
elsif ( $action eq "compile" ){
    $sql = $obj->compile;
}
elsif ( $action eq "show_space" ){
    $sql = $obj->show_space;
}
else{
    die "\n$0 doesn't know how to '$action'.\n";
} ;

print $sql;

# $Log: ddl.pl,v $
# Revision 1.1  2011/03/25 14:36:26  ekicmust
# #147852 initial version. Instead of senora.exe the perl version will be taken now.
#
# Revision 1.1.1.1  2006/05/05 20:13:35  martin
#
# initial import
#
# Revision 1.1  2004/02/12 10:57:57  DrauMart
# Initial Checkin
#
# Revision 1.13  2001/04/28 13:50:25  rvsutherland
# Modified to facilitate the new type 'schema'.
#
# Revision 1.12  2001/03/31 18:27:42  rvsutherland
# Facilitated new object type 'components', which requires neither
# name nor owner.
#
# Revision 1.11  2001/03/20 01:49:51  rvsutherland
# Facilitated instance method 'show_space'
#
# Revision 1.10  2001/03/03 18:41:31  rvsutherland
# Added DESCRIPTION to pod.
#
# Revision 1.9  2001/01/27 16:21:44  rvsutherland
# Added NAME section to pod.
#
# Revision 1.8  2001/01/14 16:47:55  rvsutherland
# Nominal changes for version 0.32
#
# Revision 1.7  2001/01/07 16:43:56  rvsutherland
# Added COPYRIGHT
#
# Revision 1.6  2001/01/06 16:21:15  rvsutherland
# Facilitated 'compile' method
#
# Revision 1.5  2000/12/09 17:55:20  rvsutherland
# Re-added after CVS bug fixed.
#
# Revision 1.3  2000/11/11 07:48:59  rvsutherland
# Added CVS tags
#

=head1 NAME

ddl.pl - Generates DDL for a single, named object

=head1 DESCRIPTION

Calls DDL::Oracle for the DDL of a specified object.

=head1 AUTHOR

 Richard V. Sutherland
 rvsutherland@yahoo.com

=head1 COPYRIGHT

Copyright (c) 2000, 2001 Richard V. Sutherland.  All rights reserved.
This module is free software.  It may be used, redistributed, and/or
modified under the same terms as Perl itself.  See:

    http://www.perl.com/perl/misc/Artistic.html

=cut

